src/pages/workspaces/workspace/DeleteWorkspaceModal.js

```diff
@@ -1,23 +1,70 @@
 import { ButtonPrimary, Link, spinnerOverlay } from 'src/components/common'
+import { warningBoxStyle } from 'src/components/data/data-utils'
+import { icon } from 'src/components/icons'
 import { TextInput } from 'src/components/input'
 import Modal from 'src/components/Modal'
 import { Ajax } from 'src/libs/ajax'
-import { bucketBrowserUrl } from 'src/libs/auth'
+import { bucketBrowserUrl, getUser } from 'src/libs/auth'
+import colors from 'src/libs/colors'
 import { reportError } from 'src/libs/error'
 import * as Utils from 'src/libs/utils'
+import { isAppDeletable } from 'src/libs/runtime-utils'
 
 
+const LoadApps = workspaceName => {
+  const signal = Utils.useCancellation()
+  const [deletableApps, setDeletableApps] = useState()
+  const [nonDeletableApps, setNonDeletableApps] = useState()
+
+  const load = async () => {
+    const [currentWorkspaceAppList] = await Promise.all([
+      Ajax(signal).Apps.listWithoutProject({ creator: getUser().email, saturnWorkspaceName: workspaceName })
+    ])
+    const appPartition = _.partition(isAppDeletable, currentWorkspaceAppList)
+    setDeletableApps(appPartition[0])
+    setNonDeletableApps(appPartition[1])
+  }
+
+  Utils.useOnMount(() => {
+    load()
+  })
+
+  return { deletableApps, nonDeletableApps }
+}
+
 const DeleteWorkspaceModal = ({ workspace: { workspace: { namespace, name, bucketName } }, onDismiss, onSuccess }) => {
   const [deleting, setDeleting] = useState(false)
   const [deleteConfirmation, setDeleteConfirmation] = useState('')
+  const { deletableApps, nonDeletableApps } = LoadApps(name)
+
+  const getAppDeletionMessage = (deletableApps, nonDeletableApps) => {
+    return nonDeletableApps.length > 0 ?
+      div({ style: { ...warningBoxStyle, fontSize: 14, display: 'flex', flexDirection: 'column' } }, [
+          div({ style: { display: 'flex', flexDirection: 'row', alignItems: 'center' } }, [
+            icon('warning-standard', { size: 19, style: { color: colors.warning(), flex: 'none', marginRight: '0.5rem' } }),
+            'Undeletable Workspace Warning'
+          ]),
+          p({ style: { fontWeight: 'normal' } }, `You cannot delete this workspace because there are ${nonDeletableApps.length} application(s) you must delete first. Only applications in ('ERROR', 'RUNNING') status can be automatically deleted.`)
+      ]) :
+      p({ style: { marginLeft: '1rem', fontWeight: 'bold' } }, `Detected ${deletableApps.length} automatically deletable application(s).`)
+  }
+
+  const hasApps = (deletableApps, nonDeletableApps) => {
+    return deletableApps !== undefined && nonDeletableApps !== undefined &&
+      (deletableApps.length > 0 ||
+      nonDeletableApps.length > 0)
+  }
 
   const deleteWorkspace = async () => {
     try {
       setDeleting(true)
       await Ajax().Workspaces.workspace(namespace, name).delete()
+      await Promise.all(
+        _.map(async app => await Ajax().Apps.app(app.googleProject, app.appName).delete(), deletableApps)
```

Comment:
this maps each deletable app to an async function which `await`s deletion before exitting. This can then be nested in Promise.all(), which must itself be `await`ed to ensure all async functions exit before continueing.